src/lib/tls/tls13/msg_certificate_13.cpp

```diff
@@ -97,28 +118,86 @@ void Certificate_13::verify(Callbacks& callbacks,
    if(!certificate_allows_signing(server_cert))
       {
       throw TLS_Exception(Alert::BAD_CERTIFICATE,
-         "Certificate usage constraints do not allow signing");
+                          "Certificate usage constraints do not allow signing");
       }
 
+   const auto trusted_CAs = creds.trusted_certificate_authorities(
+                                     m_side == Connection_Side::CLIENT ? "tls-client" : "tls-server",
+                                     hostname);
+
    const auto usage = (m_side == CLIENT) ? Usage_Type::TLS_CLIENT_AUTH : Usage_Type::TLS_SERVER_AUTH;
    callbacks.tls_verify_cert_chain(certs, ocsp_responses, trusted_CAs, usage, hostname, policy);
    }
 
-Certificate_13::Certificate_13(const std::vector<X509_Certificate>& certs,
-                               const Connection_Side side,
-                               const std::vector<uint8_t> &request_context,
-                               Callbacks& callbacks)
-   : m_request_context(request_context)
-   , m_side(side)
+void Certificate_13::setup_entries(std::vector<X509_Certificate> cert_chain,
+                                   Callbacks& callbacks)
    {
-   for (const auto& c : certs)
+   for(size_t i = 0; i < cert_chain.size(); ++i)
       {
       auto exts = Extensions();
-      callbacks.tls_modify_extensions(exts, side, type());
-      m_entries.emplace_back(Certificate_Entry{c, std::move(exts)});
+      // This will call the modification callback multiple times. Once for
+      // each certificate in the `cert_chain`. Users that want to add an
+      // extension to a specific Certificate Entry might have a hard time
+      // to distinguish them.
```

Comment:
We could perhaps add a variant that additionally accepts the certificate in question, which by default just forwards the call to `tls_modify_extensions` and discards the cert. This could happen later on, even post-3.0, so I don't consider it at all essential.

Reply:
That's roughly what I had in mind writing this comment. I'd add it when the need comes up.